### DISCUSSION TITLE
FoldWhile/FoldUntil with an async predicate

### DISCUSSION THREAD

**Discussion by fjmorel** [louthy/language-ext#1031] [Q&A] [answered]
I have a a list of strings. I want to iterate it to get an `EitherAsync<ErrorModel, DataModel>` for each string, but stop at the first entry that returns an `ErrorModel` and return that error instead of continuing.

Initially, I had this which returns what I want (first error or full list if successful) but still iterates over the entire list when there's an error:
```csharp
public EitherAsync<ErrorModel, Seq<DataModel>> GetManyAsync(Seq<string> guids)
{
    EitherAsync<ErrorModel, Seq<DataModel>> seed = new Seq<DataModel>();
    return guids
        .Aggregate(seed, (either, guid) => either
            .Bind(results => GetDataAsync(guid)
                .Map(result => results.Add(result))));
}
public EitherAsync<ServiceError, DataModel> GetDataAsync(string guid) { ... }
```

I then tried to use this library's FoldWhile/FoldUntil to stop when the Either goes Left, but I end up having to use `Task.Result` which I definitely do not want to do. 
```csharp
public EitherAsync<ErrorModel, Seq<DataModel>> GetManyAsync(Seq<string> guids)
{
    EitherAsync<ErrorModel, Seq<DataModel>> state = new Seq<DataModel>();
    return guids
        .FoldWhile(state, (either, guid) => either
            .Bind(results => GetDataAsync(guid)
                .Map(result => results.Add(result))), x => x.IsRight.Result);
}
public EitherAsync<ServiceError, DataModel> GetDataAsync(string guid) { ... }
```

Is there some other way to do this that I am missing? The original `Aggregate` isn't a huge problem because the passed in list is going to be at most a few hundred strings.

**Comment by louthy** [comment, marked as answer]
You can use `SequenceSerial` and `TraverseSerial` to turn a sequence of asynchronous actions into a asynchronous action of sequence.  

I haven't got a debugger to hand, but try this:

```csharp
    guids.SequenceSerial(GetDataAsync);
```

It will only return a sequence if all items succeed, otherwise you get the first error.  There's also `SequenceParallel` and `TraverseParallel`.  The rules are the same, but the error you get back might not be the first sequentially.

**Reply by fjmorel** [reply]
Thank you! That was exactly what I was looking for.
